File: count_triplets_form_arr_xor.py
# 1442. Count Triplets That Can Form Two Arrays of Equal XOR

"""
given an array of integers 'arr'. we want to select three indices 'i', 'j',
and 'k' where '(0 <= i < j <= k < arr.length)'. lets define 'a' and 'b' as
follows. 'a = arr[i] ^ arr[i + 1] ^ ... ^ arr[j - 1]', 'b = arr[j] ^ arr[j +
1] ^ ... ^ arr[k]'. note that '^' denotes the bitwise xor operation. return
the number of triplets (i, j, and k) where 'a == b'
"""


class Solution(object):
    def countTriplets(self, arr):
        """
        :type arr: List[int]
        :rtype: int
        """
        arr.insert(0, 0)
        n, ans = len(arr), 0
        for i in range(n - 1):
            arr[i + 1] ^= arr[i]
        for i in range(n):
            for j in range(i + 1, n):
                if arr[i] == arr[j]:
                    ans += j - i - 1
        return ans


if __name__ == "__main__":
    obj = Solution()
    print(obj.countTriplets([2, 3, 1, 6, 7]))
    print(obj.countTriplets([1, 1, 1, 1, 1]))

File: count_triplets_form_arr_xor.c
// 1442. Count Triplets That Can Form Two Arrays of Equal XOR
#include "leetcode.h"

/*
 * given an array of integers 'arr'. we want to select three indices 'i', 'j',
 * and 'k' where '(0 <= i < j <= k < arr.length)'. lets define 'a' and 'b' as
 * follows. 'a = arr[i] ^ arr[i + 1] ^ ... ^ arr[j - 1]', 'b = arr[j] ^ arr[j +
 * 1] ^ ... ^ arr[k]'. note that '^' denotes the bitwise xor operation. return
 * the number of triplets (i, j, and k) where 'a == b'
 */

int countTriplets(int *arr, int arrSize) {
  if (arrSize < 2)
    return 0;
  int ans = 0;
  for (int i = 0; i < arrSize; i++) {
    int tmp = arr[i];
    for (int j = i + 1; j < arrSize; j++) {
      tmp ^= arr[j];
      if (!tmp)
        ans += j - i;
    }
  }
  return ans;
}

int main() {
  int a1[] = {2, 3, 1, 6, 7}, a2[] = {1, 1, 1, 1, 1};
  printf("%d\n", countTriplets(a1, ARRAY_SIZE(a1))); // expect: 4
  printf("%d\n", countTriplets(a2, ARRAY_SIZE(a2))); // expect: 10
}
